nexus/src/app/webhook.rs

```diff
@@ -0,0 +1,1030 @@
+//! database schema version as well.  This way, we can detect whether a glob's
+//! list of subscriptions are up to date.  The `webhook_dispatcher` background
+//! task will query the database for any globs which were last reprocessed at
+//! earlier database schema versions and reprocess those globs prior to
+//! attempting to dispatch events to receivers.
+//!
+//! ## Deliveries, Delivery Attempts, and Liveness Probes
+//!
+//! A *delivery* represents the process of sending HTTP request(s) representing
+//! a webhook event to a receiver.  Failed HTTP requests are retried up to two
+//! times times, so a delivery may consist of up to three *delivery attempts*.
+//! Each time the `webhook_deliverator` background task is activated, it
+//! searches for deliveries which have not yet succeeded or permanently failed,
+//! hich are not presently being delivered by another Nexus, and for which the
+//! backoff period for any prior failed delivery attempts has elapsed.  It then
+//! sends an HTTP request to the webhook receiver, and records the result,
+//! creating a new `webhook_delivery_attempt` record and updating the
+//! `webhook_delivery` record.
+//!
+//! Multiple Nexii use an advisory lease mechanism to avoid attempting to
+//! deliver the same event simultaneously, by setting their UUID and a timestamp
+//! on the `webhook_delivery` record.  Because webhook delivery is
+//! at-least-once, this lease mechanism is NOT REQUIRED FOR CORRECTNESS IN ANY
+//! WAY, Andrew. :) Instead, it serves only to reduce duplicate work.
+//! Therefore, should a Nexus acquire a lease on a delivery and fail to either
+//! complete the delivery attempt within a period of time, another Nexus is
+//! permitted to clobber its lease.
+//!
+//! Deliveries are created either because an event occurred and a webhook
+//! receiver is subscribed to it, or because we were asked to resend a previous
+//! delivery that failed permanently by exhausting its retry budget.  Initial
+//! deliveries are created by activations of the webhook dispatcher background
+//! task.  When creating a delivery, the data associated with the event record
+//! in the `webhook_event` table is processed to produce the data payload that
+//! will actually be sent to the receiver. Data which the receiver's service
+//! account is not authorized to read is filtered out of the payload.[^2]
+//!
+//! Re-delivery of an event can be requested either via the event resend API
+//! endpoint, or by a *liveness probe* succeeding.  Liveness probes are
+//! synthetic delivery requests sent to a webhook receiver to check whether it's
+//! actually able to receive an event.  They are triggered via the
+//! [`Nexus::webhook_receiver_probe`] API endpoint.  A probe may optionally
+//! request that any events for which all past deliveries have failed be resent
+//! if it succeeds.  Delivery records are also created to represent the outcome
+//! of a probe.
```

Comment:
Since liveness probes appear to be bucked in the "Event class" enum as the string `probe`, what happens if someone tries to make a receiver for the probe event directly?

Do we (/should we) have an integration test for this?

Reply:
Good call, I'll add a test for that!

Reply:
Done in e85868b7858ec7040b17e1338239076e404df24a!